api/src/main/java/io/druid/data/input/impl/PrefetchableTextFilesFirehoseFactory.java

```diff
@@ -0,0 +1,400 @@
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * PrefetchableTextFilesFirehoseFactory is an abstract firehose factory for reading text files.  The firehose returned
+ * by this class provides three key functionalities.
+ *
+ * <ul>
+ * <li>Caching: for the first call of {@link #connect(StringInputRowParser, File)}, it caches objects in a local disk
+ * up to {@link #maxCacheCapacityBytes}.  These caches are NOT deleted until the process terminates,
+ * and thus can be used for future reads.</li>
+ * <li>Fetching: when it reads all cached data, it fetches remaining objects into a local disk and reads data from
+ * them.  For the performance reason, prefetch technique is used, that is, when the size of remaining cached or
+ * fetched data is smaller than {@link #prefetchTriggerBytes}, a background prefetch thread automatically starts to
+ * fetch remaining objects.</li>
+ * <li>Retry: if an exception occurs while downloading an object, it retries again up to
+ * {@link #maxFetchRetry}.</li>
+ * </ul>
+ *
+ * This implementation can be useful when the cost for reading input objects is large as reading from AWS S3 because
+ * IndexTask can read the whole data twice for determining partition specs and generating segments if the intervals of
+ * GranularitySpec is not specified.
+ */
+public abstract class PrefetchableTextFilesFirehoseFactory<ObjectType>
+    extends AbstractTextFilesFirehoseFactory<ObjectType>
+{
+  private static final Logger LOG = new Logger(PrefetchableTextFilesFirehoseFactory.class);
+  private static final long DEFAULT_MAX_CACHE_CAPACITY_BYTES = 1024 * 1024 * 1024; // 1GB
+  private static final long DEFAULT_MAX_FETCH_CAPACITY_BYTES = 1024 * 1024 * 1024; // 1GB
+  private static final long DEFAULT_FETCH_TIMEOUT = 60_000; // 60 secs
+  private static final int DEFAULT_MAX_FETCH_RETRY = 3;
+  private static final String CACHE_FILE_PREFIX = "cache-";
+  private static final String FETCH_FILE_PREFIX = "fetch-";
+
+  // The below two variables are roughly the max size of total cached/fetched objects, but the actual cached/fetched
+  // size can be larger. The reason is our current client implementations for cloud storages like s3 don't support range
+  // scan yet, so we must download the whole file at once. It's still possible for the size of cached/fetched data to
+  // not exceed these variables by estimating the after-fetch size, but it makes us consider the case when any files
+  // cannot be fetched due to their large size, which makes the implementation complicated.
+  private final long maxCacheCapacityBytes;
```

Comment:
@dclim thanks for your comment. The idea behind separating cache and fetch pools is that caching and fetching have different goals. Caching is for reducing the initial latency of the firehose while prefetching is for increasing the throughput of the firehose (it is assumed that reading directly from the remote site is slow). So, I thought cached objects should be the first n objects (it's also assumed that objects should be read in the given order and the order must not be changed if firehose is reconnected). 
Your idea about early ingestion while caching sounds good. I updated the patch.